mypy/checker.py

```diff
@@ -8010,14 +8012,26 @@ def conditional_types(
                 current_type = try_expanding_sum_type_to_union(current_type, enum_name)
         proposed_items = [type_range.item for type_range in proposed_type_ranges]
         proposed_type = make_simplified_union(proposed_items)
-        if isinstance(proposed_type, AnyType):
+        current_type = get_proper_type(current_type)
+        if isinstance(current_type, AnyType):
+            return proposed_type, current_type
+        elif isinstance(proposed_type, AnyType):
             # We don't really know much about the proposed type, so we shouldn't
             # attempt to narrow anything. Instead, we broaden the expr to Any to
             # avoid false positives
             return proposed_type, default
         elif not any(
             type_range.is_upper_bound for type_range in proposed_type_ranges
-        ) and is_proper_subtype(current_type, proposed_type, ignore_promotions=True):
+        ) and (  # concrete subtypes
+            is_proper_subtype(current_type, proposed_type, ignore_promotions=True)
+            or (  # structural subtypes
+                is_subtype(current_type, proposed_type, ignore_promotions=True)
+                and (
+                    isinstance(proposed_type, CallableType)
+                    or (isinstance(proposed_type, Instance) and proposed_type.type.is_protocol)
+                )
+            )
+        ):
```

Comment:
As a micro-optimization, let's swap the checks: `isinstance` is instant while `is_subtype` is rather heavy

```suggestion
        ) and (
            # concrete subtypes
            is_proper_subtype(current_type, proposed_type, ignore_promotions=True)
            # structural subtypes
            or (
                (
                    isinstance(proposed_type, CallableType)
                    or (isinstance(proposed_type, Instance) and proposed_type.type.is_protocol)
                )
                and is_subtype(current_type, proposed_type, ignore_promotions=True)
            )
        ):
```